Add backing traits for all clients

mthjones opened the pull request:
## Description
This PR modifies the codegen to generate a trait for every client that defines the functions the client has to implement. For example, the existing `DynamoDbClient<P,D>` now implements a `DynamoDb` trait.

## Motivation
The purpose of adding these traits is two-fold:
1. it resolves the ergonomics issue brought up in #619 where everything that wants to use or store a client must now be generic over `<P,D>` as well
2. it allows users to create mock versions of clients to verify in their testing without having to drop down to mocking the dispatcher itself

## Notes
* This is similar to what is done in official Amazon SDKs like the .NET one, where they provide both an `IAmazonDynamoDb` and an `AmazonDynamoDbClient`.
* This is a **BREAKING CHANGE**. Users will have to import the trait instead of the client in most cases.
* I haven't run the integration tests against an AWS account, because I don't believe there should be any functional changes here that would not be a compilation error. Trying to compile with `all` features brings my computer to its knees, but I've verified `dynamodb`, `s3` and `sts`. `sts` in particular, because it has some unique code.

matthewkmayer commented:
Thanks for the PR!

At first glance I'm uncertain if this is functionality we want in Rusoto or `rusoto_helpers`.  The `rusoto_helpers` crate has languished for a while but is meant to provide higher level abstractions.  Less fussing around with service client instances and `hyper` and all that.  Not to say we shouldn't have this in the crate now and possibly split later, just explaining why this hasn't come up yet.

I'll review this soon. 😄

mthjones commented:
I don't believe this could actually go in `rusoto_helpers`, unless the client was implemented in the `impl` and the trait implementation just called the `impl` implementation... or `rusoto` were to have a dependency on `rusoto_helpers` ( circular dependencies... 😬 ).

I don't think that this is a higher-level abstraction, like [this](https://github.com/mthjones/pierre/blob/dynamodb_backend/daemon/src/store/dynamodb.rs), it's just a way to be able to work with clients without exposing implementation details (`hyper::Client` everywhere) or infecting your codebase with `<P,D> ... where P: rusoto::ProvideAwsCredentials, D: DispatchSignedRequest` everywhere.

My understanding is that the helpers are manually-implemented, which would negate the ability to codegen these traits at the same time as the clients.

matthewkmayer commented:
I believe you're correct on all counts. 😄 

Yes, helpers were manually generated and are probably so far out of date we may want to remove them from the repo entirely until we get around to them.  I don't want them to mislead people.

I'm giving the PR a good once-over.  One thing I'd like to see is updated docs on how to use the `trait` instead of the classes directly, showing off this great new behavior.  This could go in a different PR to keep this one from getting too huge.

matthewkmayer commented on tests/elasticache.rs:
```diff
@@ -2,7 +2,7 @@
 extern crate rusoto;
 
-use rusoto::elasticache::{ElastiCacheClient, DescribeCacheClustersMessage};
+use rusoto::elasticache::{Elasticache, ElastiCacheClient, DescribeCacheClustersMessage};
```
The new trait import should be `ElastiCache`.

```bash
error[E0432]: unresolved import `rusoto::elasticache::Elasticache`
 --> tests/elasticache.rs:5:27
  |
5 | use rusoto::elasticache::{Elasticache, ElastiCacheClient, DescribeCacheClustersMessage};
  |                           ^^^^^^^^^^^ no `Elasticache` in `elasticache`. Did you mean to use `ElastiCache`?
```

Taking that compiler suggestion makes it work. 😄

mthjones replied:
Interesting...that definitely shouldn't have gotten past CI. Good catch though!

Oh, wait, no. Only lib tests are done in CI. It would be handy to at least build integration tests to verify them though. There's a new flag for Cargo coming that allows passing `--tests` to `cargo build` to build tests, but I don't think it will be dropping until 1.19. Hrm... I guess the only good way to catch that is locally.

matthewkmayer replied:
It's an issue because [CI doesn't run tests that create real AWS resources](https://github.com/rusoto/rusoto/issues/222).  I've been working on getting an AWS account for Rusoto so CI can run them but no luck yet.  There's something new in the works that may get traction for solving it.

mthjones commented:
I agree. I don't mind adding documentation to this PR, to keep everything together. Didn't want to spend a bunch of documentation effort before veryifing that it was going to make it in. 😁

matthewkmayer commented on tests/cloudformation.rs:
```diff
@@ -2,7 +2,7 @@
 extern crate rusoto;
 
-use rusoto::cloudformation::{CloudFormationClient, ListStacksInput};
+use rusoto::cloudformation::{Cloudformation, CloudFormationClient, ListStacksInput};
```
Another casing hiccup: new text should be `CloudFormation`.

matthewkmayer commented:
Aside from the casing changes for integration tests this looks great, thank you.

Once those are taken care of and docs are added this is good to go. 🎉

mthjones commented:
Thanks for the review! Will get docs and those two fixes up ASAP. Just give me a bit, I built "packagea" and my computer locked up. 😄

matthewkmayer commented:
Crategen is very high on the priority list to alleviate this. 😄

mthjones commented:
Added a link to the trait at the top of every module, and added a generated doc comment on the traits. Examples are already updated. The compiler error message is pretty helpful, and I think as long as the change is listed in the release notes, it should be fine.

drusellers commented:
Yay

matthewkmayer commented:
Thanks!
